gnovm/pkg/gnolang/type_check.go

```diff
@@ -0,0 +1,779 @@
+
+	var xt, dt Type
+	cmp := cmpSpecificity(lt, rt) // check potential direction of type conversion
+	if cmp <= 0 {
+		xt = lt
+		dt = rt
+	} else {
+		xt = rt
+		dt = lt
+	}
+
+	if isComparison(bx.Op) {
+		switch bx.Op {
+		case EQL, NEQ:
+			assertComparable(xt, dt) // only check if dest type is comparable
+		case LSS, LEQ, GTR, GEQ:
+			if checker, ok := binaryChecker[bx.Op]; ok {
+				bx.checkCompatibility(store, xt, dt, checker, escapedOpStr)
+			} else {
+				panic("should not happen")
+			}
+		default:
+			panic("invalid comparison operator")
+		}
+	} else {
+		if checker, ok := binaryChecker[bx.Op]; ok {
+			bx.checkCompatibility(store, xt, dt, checker, escapedOpStr)
+		} else {
+			panic("should not happen")
+		}
+
+		switch bx.Op {
+		case QUO, REM:
+			// special case of zero divisor
+			if isQuoOrRem(bx.Op) {
+				if rcx, ok := bx.Right.(*ConstExpr); ok {
+					if rcx.TypedValue.isZero() {
+						panic("invalid operation: division by zero")
+					}
+				}
+			}
+		default:
+			// do nothing
+		}
+	}
+}
+
+func (bx *BinaryExpr) checkCompatibility(store Store, xt, dt Type, checker func(t Type) bool, escapedOpStr string) {
+	var destKind interface{}
+
+	if !checker(dt) {
+		if dt != nil {
+			destKind = dt.Kind()
+		}
+		panic(fmt.Sprintf("operator %s not defined on: %v", escapedOpStr, destKind))
+	}
+
+	defer func() { // rewrite err msg
+		if r := recover(); r != nil {
+			if xt != nil {
+				destKind = xt.Kind()
+			}
+			panic(fmt.Sprintf("operator %s not defined on: %v", escapedOpStr, destKind))
+		}
+	}()
+
+	// e.g. 1%1e9
+	if !checker(xt) {
+		checkAssignableTo(xt, dt, false) // XXX, cache this?
+	}
+}
+
+func (ux *UnaryExpr) AssertCompatible(xt, dt Type) {
+	var destKind interface{}
+
+	if nt, ok := xt.(*NativeType); ok {
+		if _, ok := go2GnoBaseType(nt.Type).(PrimitiveType); ok {
+			return
+		}
+	}
+	// check compatible
+	if checker, ok := unaryChecker[ux.Op]; ok {
+		if dt == nil {
+			dt = xt
+		}
+		if !checker(dt) {
+			if dt != nil {
+				destKind = dt.Kind()
+			}
+			panic(fmt.Sprintf("operator %s not defined on: %v", wordTokenStrings[ux.Op], destKind))
+		}
+	} else {
+		panic("should not happen")
+	}
+}
+
+func (idst *IncDecStmt) AssertCompatible(t Type) {
+	var destKind interface{}
+
+	if nt, ok := t.(*NativeType); ok {
+		if _, ok := go2GnoBaseType(nt.Type).(PrimitiveType); ok {
+			return
+		}
+	}
+	// check compatible
+	if checker, ok := IncDecStmtChecker[idst.Op]; ok {
+		if !checker(t) {
+			if t != nil {
+				destKind = t.Kind()
+			}
+			panic(fmt.Sprintf("operator %s not defined on: %v", wordTokenStrings[idst.Op], destKind))
+		}
+	} else {
+		panic("should not happen")
+	}
+}
+
+func (as *AssignStmt) AssertCompatible(store Store, last BlockNode) {
+	escapedOpStr := strings.Replace(wordTokenStrings[as.Op], "%", "%%", 1)
+	var destKind interface{}
+
+	// XXX, assume lhs length is same with of rhs
+	// Call case: a, b = x(...)
+	for i, x := range as.Lhs {
+		lt := evalStaticTypeOf(store, last, x)
```

Comment:
I think there needs to be something after this to avoid passing for cases like, `1 = 6, false = true, "hello" = "world", etc`. Maybe something like:
```go
if _, ok := x.(*NameExpr); !ok && isUntyped(lt) {
	panic("lhs of assigment must be named")
}
```

Reply:
yes there lacks some check, also with this:
also this one is not checked either:
package main

func foo() (int, float32) {
	return 1, 1.0
}

func main() {
	var s, s1 string
	s, s1 = foo()
	println(s)
}